src/services/ModifyFile.ts

```diff
@@ -66,16 +70,25 @@ export class ModifyFile {
     return this;
   }
 
-  putFileFromBytes(path: string, bytes: Buffer) {
-    let end = GRPC_MAX_MESSAGE_LENGTH;
+  putFileFromBytes(path: string, bytes: Buffer, callback?: () => void) {
+    const messageLength = GRPC_MAX_MESSAGE_LENGTH - STREAM_OVERHEAD_LENGTH;
+    let end = messageLength;
     let chunk = bytes.slice(0, end);
-    while (chunk.length > 0) {
-      const addFile = new AddFile()
-        .setPath(path)
-        .setRaw(new BytesValue().setValue(chunk));
-      this.stream.write(new ModifyFileRequest().setAddFile(addFile));
-      chunk = bytes.slice(end, end + GRPC_MAX_MESSAGE_LENGTH);
-      end += GRPC_MAX_MESSAGE_LENGTH;
+    const write = () => {
+      let ok = true;
+      while (chunk.length > 0 && ok) {
+        const addFile = new AddFile()
+          .setPath(path)
+          .setRaw(new BytesValue().setValue(chunk));
+        ok = this.stream.write(new ModifyFileRequest().setAddFile(addFile));
+        chunk = bytes.slice(end, end + messageLength);
+        end += messageLength;
+        if (chunk.length <= 0 && callback) callback();
+      }
+    };
+    write();
+    if (chunk.length > 0) {
+      this.stream.once('drain', write);
```

Comment:
Just double checking my understanding - this is to resume writing to the stream once `write` is in a `!ok` state?

It might be worth just making a note about the `'drain'` event here.